### DISCUSSION TITLE
Expose useFormContext

### DISCUSSION THREAD

**Discussion by UrbanChrisy** [TanStack/form#600] [Ideas] [answered]
I want to build build a mutli-stage form, but currently at the moment I need to pass the formApi object from the `useForm` hook down to all my components or create a a provider myself. When then is duplicating the work.

Would be good if the api `useFormContext` is exposed for us to use, considering its already being used in useField. I think it would be a one or two line change, just exporting the file.

**Comment by UrbanChrisy** [comment]
Usage in `useField`
https://github.com/TanStack/form/blob/eb4801c5202cbe3007401047f3c776351b0631b4/packages/react-form/src/useField.tsx#L68

**Comment by UrbanChrisy** [comment]
PR made: https://github.com/TanStack/form/pull/601

**Comment by crutchcorn** [comment, marked as answer]
This is no longer relevant as we don't have a context anymore

**Reply by devth** [reply]
How can a child react component get a ref to the form in a parent component?

**Reply by crutchcorn** [reply]
Pass it as a prop or make a context yourself

**Reply by arkadiy93** [reply]
@crutchcorn How would you type the form in the context provider?
I tried using ReactFormExtendedApi but it requires 10 args(!)

edit: Ended up inferring return type of function that creates the form. Would be nice if we can get an exported type that accepts just the form type as an arg.

**Reply by nijikokun** [reply]
I upvoted @arkadiy93 but really would like this as well.

**Reply by martin-dimi** [reply]
Same here. At least exposing a nice to work with type would be nice

**Reply by edwardAgileLab** [reply]
up

**Reply by itzfeltrin** [reply]
up

**Reply by ben-heimberg** [reply]
@arkadiy93 
Can you give an example?

**Reply by arkadiy93** [reply]
@ben-heimberg 
Create a hook that creates and returns the form, lets say useCreateForm -

`type FormType = ReturnType<typeof useCreateForm>;`

**Reply by Balastrong** [reply]
You can use the [Form Composition APIs](https://tanstack.com/form/latest/docs/framework/react/guides/form-composition) to create forms with shared context, you don't need to do that manually.

**Reply by ben-heimberg** [reply]
@Balastrong Yes, but that still doesnt give access to all form values/errors from what i understand. I can use 'withForm' but thats still prop drilling.

I need access to form values and meta data pretty deep in the component tree.

Or have i misunderstood something?

**Reply by acropolis11** [reply]
Have we found a way to share the form through context with type safety ?

**Reply by colorpulse6** [reply]
Also looking for a coherent solution to this

**Reply by romankoho** [reply]
me as well

**Comment by gnesher** [comment]
@crutchcorn the form composition guide (https://tanstack.com/form/latest/docs/framework/react/guides/form-composition#pre-bound-field-components) still uses `useFormContext` which is leading to some confusion - I assume this is just outdated and should not be used?

**Reply by ccrutchley-ih** [reply]
Let's talk a little about the history of TanStack Form:

- It originally had a form-wide context provided for all forms by default
    - This made large-forms easier, but stripped type safety and other problems
- As a result, we figured out a way to remove context (hence my "We have no context anymore")
- Since then, however, we've re-added the context for a different purpose: Generic component usage. This is the most up-to-date and accurate way of writing Form code.

Sorry for the confusion - you sort've just have stumbled on historical records here.

**Reply by colorpulse6** [reply]
I feel that i am testing the limits of form composition and reusability and haven't used `useFormContext` at all because the form we get there is not the same type as the form from the hook. I have however used `useFieldContext` with a generic as a `fieldComponent`

**Reply by ccrutchley-ih** [reply]
@colorpulse6 why don't you join our Discord and discuss more there? I suspect there's some misunderstandings with your usage of Form based on what you're saying.

**Reply by acropolis11** [reply]
Just to confirm, if I have a form whose values and methods I want to access from a component 3 levels down the component tree. I would have to wrap each child component in the withForm HOC ?

just curious if this is the best approach to handle this.
